Time Stepping tutorials

valeriabarra opened the pull request:
### Description

This PR adds some Time Stepping tutorials. The aim is for the users/learners to learn about the different Time Steppers/Integrators available in the ClimateMachine and their different properties and usage.

- [x] Code follows the [style guidelines](https://clima.github.io/ClimateMachine.jl/latest/DevDocs/CodeStyle/) OR N/A.
- [x] Unit tests are included OR N/A.
- [x] Code is exercised in an integration test OR N/A.
- [x] Documentation has been added/updated OR N/A.

valeriabarra commented:
Opened so we can preview/test via `bors try` but converted to draft for now

valeriabarra commented:
bors try

valeriabarra commented:
bors try

valeriabarra commented:
bors try

valeriabarra commented:
bors try

valeriabarra commented:
bors try-

valeriabarra commented:
bors try

valeriabarra commented:
bors try-

valeriabarra commented:
bors try

valeriabarra commented:
bors try

valeriabarra commented:
bors try-

valeriabarra commented:
bors try

valeriabarra commented:
bors try

valeriabarra commented:
bors try

valeriabarra commented:
bors try

valeriabarra commented:
bors try

valeriabarra commented:
bors try

valeriabarra commented:
bors try

thomasgibson commented:
bors try

thomasgibson commented:
bors try

valeriabarra commented:
bors try

valeriabarra commented:
bors try-

valeriabarra commented:
bors try

simonbyrne commented on docs/list_of_tutorials.jl:
```diff
@@ -60,6 +60,11 @@ if generate_tutorials
             "DG Methods" =>
                 ["Filters" => "Numerics/DGMethods/showcase_filters.jl"],
+            "Time-Stepping" => [
+                "Introduction" => "Numerics/TimeStepping/ts_intro.jl",
```
I think it needs to be added here:
```suggestion
                "Introduction" => "Numerics/TimeStepping/ts_intro.jl",
                "Numerics/TimeStepping/tutorial_risingbubble_config.jl",
```

valeriabarra replied:
Makes sense, thank you!

valeriabarra commented:
bors try

valeriabarra commented:
bors try

valeriabarra commented:
bors try

simonbyrne commented on tutorials/Numerics/TimeStepping/tutorial_risingbubble_config.jl:
```diff
@@ -0,0 +1,316 @@
+#md #     `Keywords` are used to specify some arguments (see appropriate source
+#md #     files).
+
+# ## [Diagnostics](@id config_diagnostics)
```
It seems to duplicate a label elsewhere. Unless you need to refer to it, you can leave the id empty
```suggestion
# ## Diagnostics
```

valeriabarra commented:
bors try

valeriabarra commented:
bors try

valeriabarra commented:
bors try

simonbyrne commented on tutorials/Numerics/TimeStepping/tutorial_risingbubble_config.jl:
```diff
@@ -0,0 +1,316 @@
+    ## Assign State Variables
+    state.ρ = ρ
+    state.ρu = ρu
+    state.ρe = ρe_tot
```
This has changed on master. @charleskawczynski what should be done here?

charleskawczynski replied:
`state.energy.ρe = ρe_tot`

valeriabarra replied:
Thanks!

charleskawczynski commented on tutorials/Numerics/TimeStepping/tutorial_acousticwave_config.jl:
```diff
@@ -0,0 +1,104 @@
+
+    state.ρ = air_density(ts)
+    state.ρu = SVector{3, FT}(0, 0, 0)
+    state.ρe = state.ρ * (e_int + e_pot)
```
```suggestion
    state.energy.ρe = state.ρ * (e_int + e_pot)
```

valeriabarra replied:
Thanks!

charleskawczynski commented on tutorials/Numerics/TimeStepping/tutorial_risingbubble_config.jl:
```diff
@@ -0,0 +1,316 @@
+#md #     The following variables are assigned in the initial condition
+#md #     - `state.ρ` = Scalar quantity for initial density profile
+#md #     - `state.ρu`= 3-component vector for initial momentum profile
+#md #     - `state.ρe`= Scalar quantity for initial total-energy profile
```
```suggestion
#md #     - `state.energy.ρe`= Scalar quantity for initial total-energy profile
```

valeriabarra commented:
bors try

valeriabarra commented:
bors try-

valeriabarra commented:
bors try

charleskawczynski commented on tutorials/Numerics/TimeStepping/imex_ark.jl:
```diff
@@ -0,0 +1,18 @@
+
+# Lore ipsum
+
+include(joinpath(
+    @__DIR__,
+    "../../../../../",
+    "tutorials/Numerics/TimeStepping/tutorial_acousticwave_config.jl",
+))
```
```suggestion
using ClimateMachine
clima_dir = dirname(dirname(pathof(ClimateMachine)));
include(joinpath(clima_dir, "tutorials/Numerics/TimeStepping/tutorial_acousticwave_config.jl"))
```
I've found this to be a useful pattern :slightly_smiling_face:

valeriabarra replied:
Oh, does `clima_dir` in this case points to the generated path for `climatemachine-docs`?

charleskawczynski replied:
`clima_dir` is a string that points to `ClimateMachine.jl/`

charleskawczynski replied:
Maybe we should change the name everywhere to `climate_machine_dir` (this was originally written back when ClimateMachine.jl was CLIMA.jl). Maybe for a separate PR, though.

valeriabarra replied:
Thanks Charlie! I adopted your suggestion. Looks much neater.

valeriabarra commented:
bors try

valeriabarra commented:
bors try

valeriabarra commented:
bors try-

valeriabarra commented:
bors try

valeriabarra commented:
bors try

valeriabarra commented:
bors try

valeriabarra commented:
bors try

valeriabarra commented:
bors try

valeriabarra commented:
bors try

valeriabarra commented:
bors try

valeriabarra commented:
bors try

valeriabarra commented:
bors try-

valeriabarra commented:
bors try

jakebolewski commented:
It looks like this is still timing out, we set the limit here: https://github.com/CliMA/ClimateMachine.jl/blob/master/.buildkite/docs-pipeline.yml#L40

Some options would be to bump  the parallelism higher (maybe 8 -> 10) to see if that brings down the time.  The time these tutorials take though is getting prohibitive so probably it is time to rethink how they are built (just ensuring they work for every release for instance).

valeriabarra commented:
I agree this is becoming prohibitive. 
For the suggestion of bumping up the number of tasks, would I just change this [line](https://github.com/CliMA/ClimateMachine.jl/blob/25c50fa1d0d5b22bf3e29d09d04c6e05868f632f/.buildkite/docs-pipeline.yml#L34) ? Or `slurm_ntasks` is somewhere else?

valeriabarra commented:
I can open an issue to follow up on the conversation we had on Slack a while ago on how we could rethink our tutorials and their function.

valeriabarra commented:
Opened #2057

valeriabarra commented:
bors try

valeriabarra commented:
bors try

valeriabarra commented:
bors try

valeriabarra commented:
bors try

valeriabarra commented:
This is ready for review! 

Sorry about the embarrassingly long conversation in this PR. I needed to invoke `bors try` so many times to see previews. 

The new tutorials can be seen by clicking the last of the GitHub Actions `documenter/deploy` , by clicking on Details. The new tutorials are in `Tutorials` > `Numerics` > `Time-Stepping`. There are 4 different sections there. 

Please let me know if you have any feedback. @thomasgibson and I would appreciate it! 
When approved, I will take care of reducing the number of commits.

valeriabarra commented:
PS: Note: this PR also bumps up the number of tasks for slurm to 10. Otherwise, it would time out

charleskawczynski commented on tutorials/Numerics/TimeStepping/tutorial_risingbubble_config.jl:
```diff
@@ -0,0 +1,316 @@
+    return ClimateMachine.DiagnosticsConfiguration([dgngrp])
+end
+
+function run_simulation(ode_solver, CFL, timeend)
+    ## These are essentially arguments passed to the
+    ## [`config_risingbubble`](@ref config-helper) function.  For type
+    ## consistency we explicitly define the problem floating-precision.
+    FT = Float64
```
```suggestion
function run_simulation(ode_solver, CFL::FT, timeend::FT) where {FT}
```
We could alternatively grab it from `CFL` and `timeend`

charleskawczynski commented on tutorials/Numerics/TimeStepping/tutorial_risingbubble_config.jl:
```diff
@@ -0,0 +1,316 @@
+
+# ## [Initial Conditions](@id init-rtb)
+# This example demonstrates the use of functions defined
+# in the [`Thermodynamics`](@ref ClimateMachine.Thermodynamics) package to
```
```suggestion
# in the [`Thermodynamics`](@ref ClimateMachine.Thermodynamics) module to
```

charleskawczynski commented on tutorials/Numerics/TimeStepping/tutorial_risingbubble_config.jl:
```diff
@@ -0,0 +1,316 @@
+# # [Rising Thermal Bubble Configuration](@id Rising-Thermal-Bubble-Configuration)
```
Is this tutorial meant to be run? Looking at the output, it seems that all of the methods are defined, but never called

charleskawczynski replied:
Ah, nevermind, I see.

charleskawczynski commented on tutorials/Numerics/TimeStepping/tutorial_acousticwave_config.jl:
```diff
@@ -0,0 +1,112 @@
+    return nothing
+end
+
+function run_acousticwave(ode_solver, CFL, CFL_direction, timeend)
+    FT = Float64
```
We could also grab `FT` from the type system here
```suggestion
function run_acousticwave(ode_solver, CFL::FT, CFL_direction, timeend::FT) where {FT}
```

valeriabarra commented:
bors try

valeriabarra commented:
bors try

jiahe23 commented on tutorials/Numerics/TimeStepping/ts_intro.jl:
````diff
@@ -0,0 +1,35 @@
+# \end{align}
+# ```
+
+# where we have used ``\dot {\boldsymbol{q}} = d \boldsymbol{q} / dt``.
````
Could we use the regular `q` instead of `\boldsymbol{q}`?  It sits low and the doc above it is too far away.

valeriabarra replied:
It serves to express a vector quantity though. What do you mean by it sits low? It makes it bold type

jiahe23 commented on tutorials/Numerics/TimeStepping/ts_intro.jl:
```diff
@@ -0,0 +1,35 @@
+# flavor (e.g., explicit, semi-implicit, single-stage, multi-stage, single-step,
+# multi-step, single-rate, multi-rate, etc). ClimateMachine supports several
+# of them. Before showing the different nature of some of these methods, let us
+# introduce some common notation.
```
notations

jiahe23 commented on tutorials/Numerics/TimeStepping/ts_intro.jl:
````diff
@@ -0,0 +1,35 @@
+# ```
+
+# where we have used ``\dot {\boldsymbol{q}} = d \boldsymbol{q} / dt``.
+# In both explicit/implicit cases, we refer to the term ``\mathcal{G}``
````
explicit/implicit -> explicit and implicit

jiahe23 commented on tutorials/Numerics/TimeStepping/ts_intro.jl:
```diff
@@ -0,0 +1,35 @@
+
+# where we have used ``\dot {\boldsymbol{q}} = d \boldsymbol{q} / dt``.
+# In both explicit/implicit cases, we refer to the term ``\mathcal{G}``
+# as the right-hand-side (RHS) or explicit term, and to the spatial terms of
```
the spatial term

jiahe23 commented on tutorials/Numerics/TimeStepping/ts_intro.jl:
```diff
@@ -0,0 +1,35 @@
+# where we have used ``\dot {\boldsymbol{q}} = d \boldsymbol{q} / dt``.
+# In both explicit/implicit cases, we refer to the term ``\mathcal{G}``
+# as the right-hand-side (RHS) or explicit term, and to the spatial terms of
+# ``\mathcal{F}`` as the left-hand-side or implicit term.
```
left-hand-side (LHS)

szy21 commented on tutorials/Numerics/TimeStepping/explicit_lsrk.jl:
```diff
@@ -0,0 +1,190 @@
+#     D^{T} M (F^{adv} + F^{visc}) + \sum_{f=1}^{N_f} L^T M_f(\widehat{F}^{adv} + \widehat{F}^{visc})
+#     \right) \equiv \mathcal{T}(\boldsymbol{q}).
+# \end{aligned}
+# ``
```
Should we define the notations in the above equation?

valeriabarra replied:
I agree those should be explained. However perhaps it is not in the scope of this PR. We should have a section filled with the mathematical formulation chosen. Perhaps this [section](https://clima.github.io/ClimateMachine.jl/previews/PR1938/APIs/Numerics/DGMethods/DGMethods/#Mathematical-Formulation) that I am just seeing now it says "to be filled"

jiahe23 commented on tutorials/Numerics/TimeStepping/explicit_lsrk.jl:
```diff
@@ -0,0 +1,190 @@
+# kinematic viscosity. The first term on the right is the time step condition
+# due to the non-dissipative components, while the second term to the dissipation.
+# For explicit time-integrators, we have to find the minimum time step that
+# satisfies this condition along all three spatial directions.
```
Does it mean to specify the CFL_direction to be EveryDirection? If so, it will be helpful to put a note here.

jiahe23 commented on tutorials/Numerics/TimeStepping/explicit_lsrk.jl:
````diff
@@ -0,0 +1,190 @@
+# 	\boldsymbol{q}^{n+1} = \boldsymbol{q}^n + \Delta t \sum_{i=1}^{s} b_i \mathcal{T}(\boldsymbol{Q}^i),
+# \end{align}
+# ```
+#
````
I understand that the choice of \boldsymbol might aim to differentiate between var and function, but it makes it hard to read. Especially \boldsymbol{Q}^i and \boldsymbol{Q}^j, the superscript is hard to tell

valeriabarra replied:
`\boldsymbol` serves to type math symbols (including latin and greek letters) in bold type. It serves to differentiate between scalar and vector quantities, not really a var and a function. Thanks for catching the superscript! That should have not be bold. I'm going to fix it

jiahe23 commented on tutorials/Numerics/TimeStepping/explicit_lsrk.jl:
```diff
@@ -0,0 +1,190 @@
+#
+# ### [Low-storage Runge-Kutta methods](@id lsrk)
+# `ClimateMachine.jl` contains the following low-storage methods:
+#   - Forward Euler (`LSRKEulerMethod`)
```
Could we add a link to the method in the API section?

jiahe23 replied:
Same for the following two methods

jiahe23 commented on tutorials/Numerics/TimeStepping/explicit_lsrk.jl:
```diff
@@ -0,0 +1,190 @@
+# For the rising bubble example used here, we use 4th order polynomials in
+# a discontinuous Galerkin approximation, with a domain resolution of
+# 125 meters in each spatial direction. This gives an effective
+# minimanl node-distance (distance between LGL nodes) of 86 meters
```
minimal nodal distance

jiahe23 commented on tutorials/Numerics/TimeStepping/explicit_lsrk.jl:
```diff
@@ -0,0 +1,190 @@
+# 125 meters in each spatial direction. This gives an effective
+# minimanl node-distance (distance between LGL nodes) of 86 meters
+# over the entire mesh. Using the equation for the explcit time-step above,
+# we can determine the ``\Delta t`` by specifying the desired Courant number ``C``.
```
Could we change C to CFL? It is more self-explaining and is used in most of the current experiment drivers

valeriabarra replied:
Thanks for pointing that out. I thought it some other parts of the code Courant was used (see [here](https://clima.github.io/ClimateMachine.jl/previews/PR1938/APIs/Numerics/DGMethods/Courant/)) and that's how I heard most people referring to it here. I am going to rename it

szy21 commented:
Thank you @valeriabarra! This is very helpful. I wonder if it is possible to have a table listing the recommended courant number (or a range) for each time stepping method? I think it would be useful for the users.

jiahe23 commented on tutorials/Numerics/TimeStepping/explicit_lsrk.jl:
```diff
@@ -0,0 +1,190 @@
+timeend = FT(100)
+ode_solver =
+    ClimateMachine.ExplicitSolverType(solver_method = LSRK54CarpenterKennedy)
+C = FT(0.4)
```
same here and below

jiahe23 commented on tutorials/Numerics/TimeStepping/explicit_lsrk.jl:
```diff
@@ -0,0 +1,190 @@
+# `LSRK144NiegemannDiehlBusch` contains the largest stability region of the
+# low-storage methods available in `ClimateMachine.jl`.
+
+# ### [Strong Stability Preserving Runge--Kutta methods](@id ssprk)
```
Strong Stability Preserving Runge--Kutta methods (SSPRK)

jiahe23 commented on tutorials/Numerics/TimeStepping/explicit_lsrk.jl:
```diff
@@ -0,0 +1,190 @@
+# For more information on general RK methods, we refer the interested reader
+# to Ch. 5.2 of [Atkinson2011](@cite).
+#
+# ### [Low-storage Runge-Kutta methods](@id lsrk)
```
Low-storage Runge-Kutta methods (LSRK)

jiahe23 commented on tutorials/Numerics/TimeStepping/explicit_lsrk.jl:
```diff
@@ -0,0 +1,190 @@
+# ``\boldsymbol{q}^{n+1} = \boldsymbol{Q}^{(N+1)}``. This allows the updates
+# to be performed with only three copies of the state vector
+# (storing ``\boldsymbol{q}^n``, ``\boldsymbol{Q}^{i}`` and ``\mathcal{T}(\boldsymbol{Q}^{i})``).
+# We illustrate here the use of a `SSPRK33ShuOsher` method.
```
Could we link the method to its API page?

jiahe23 commented on tutorials/Numerics/TimeStepping/explicit_lsrk.jl:
```diff
@@ -0,0 +1,190 @@
+# We illustrate here the use of a `SSPRK33ShuOsher` method.
+
+ode_solver = ClimateMachine.ExplicitSolverType(solver_method = SSPRK33ShuOsher)
+C = FT(0.2)
```
same here CFL

jiahe23 commented on tutorials/Numerics/TimeStepping/imex_ark.jl:
```diff
@@ -0,0 +1,180 @@
+# for horizontally-propogating acoustic waves. We use, ``C = 0.002`` in the
+# horizontal, which corresponds to a timestep size of approximately ``1`` second.
+
+C = FT(0.002)
```
CFL

jiahe23 commented on tutorials/Numerics/TimeStepping/imex_ark.jl:
```diff
@@ -0,0 +1,180 @@
+# schemes, we start here by going over a simple, fully explicit scheme. The
+# reader can refer to the [Single-rate Explicit Timestepping tutorial](@ref Single-rate-Explicit-Timestepping)
+# for detailes on such schemes. Here we use the the 14-stage LSRK method
+# `LSRK144NiegemannDiehlBusch`, which contains the largest stability region of
```
Link the method to its API page

jiahe23 commented on tutorials/Numerics/TimeStepping/imex_ark.jl:
```diff
@@ -0,0 +1,180 @@
+# For the acoustic wave example used here, we use 4th order polynomials in
+# our discontinuous Galerkin approximation, with 6 elements in each horizontal
+# direction and 4 elements in the vertical direction, on the cubed-sphere.
+# This gives an effective minimanl node-distance (distance between LGL nodes)
```
minimal

jiahe23 commented on tutorials/Numerics/TimeStepping/imex_ark.jl:
```diff
@@ -0,0 +1,180 @@
+ode_solver = ClimateMachine.IMEXSolverType(
+    solver_method = ARK2GiraldoKellyConstantinescu,
+)
+C = FT(0.5)
```
CFL

valeriabarra replied:
Thank you, I am going to change all its occurrences

valeriabarra commented:
> Thank you @valeriabarra! This is very helpful. I wonder if it is possible to have a table listing the recommended courant number (or a range) for each time stepping method? I think it would be useful for the users.

Thank you @szy21 ! I am not sure such a table could be exhaustive. Would you think for the two examples taken here or all possible test cases? Also, for some of the methods (e.g., multirate R-K and mis) compiling such a table to find the maximum possible time step (or maximum region of stability) is actually an open research question

jiahe23 commented on tutorials/Numerics/TimeStepping/multirate_rk.jl:
````diff
@@ -0,0 +1,134 @@
+# ```
+#
+# where the coefficients ``a``, ``b``, and ``c`` correspond to the Butcher
+# tableau for a given RK method. The superscripts O`` and I`` denote the
````
``O`` and ``I`` ？

jiahe23 commented on tutorials/Numerics/TimeStepping/multirate_rk.jl:
```diff
@@ -0,0 +1,134 @@
+)
+
+timeend = FT(3600)
+C = FT(5)
```
CFL

jiahe23 commented on tutorials/Numerics/TimeStepping/multirate_rk.jl:
```diff
@@ -0,0 +1,134 @@
+C = FT(5)
+cfl_direction = HorizontalDirection()
+run_acousticwave(ode_solver, C, cfl_direction, timeend)
+
```
It outputs a number after the simulation. It is better to explain what that stands for.

valeriabarra replied:
Oh, definitely. Thank you. We actually had decided to suppress the output in the other tutorials This one must have slipped!

szy21 commented:
> Thank you @szy21 ! I am not sure such a table could be exhaustive. Would you think for the two examples taken here or all possible test cases? Also, for some of the methods (e.g., multirate R-K and mis) compiling such a table to find the maximum possible time step (or maximum region of stability) is actually an open research question

I was guessing the CFL is mostly determined by the method and is less sensitive to the experiment, but I could be wrong. I understand it could be hard, as I saw in the doc that the CFL is sometimes heuristically determined:) I was just thinking about having something so that the users would know e.g. we could use a larger CFL for `LSRK144NiegemannDiehlBusch` than for `LSRK54CarpenterKennedy`, but I don't know what would be the best way to show this.

jiahe23 commented on tutorials/Numerics/TimeStepping/mis.jl:
```diff
@@ -0,0 +1,83 @@
+# tutorial. See its [tutorial page](@ref Rising-Thermal-Bubble-Configuration)
+# for details on the model and parameters. For the purposes of this tutorial,
+# we will only run the experiment for a total of 100 simulation seconds.
+
```
Are we demonstrating it thru the bubble or the acoustic wave?

valeriabarra replied:
Woops! Thanks for catching that.

valeriabarra commented:
> > Thank you @szy21 ! I am not sure such a table could be exhaustive. Would you think for the two examples taken here or all possible test cases? Also, for some of the methods (e.g., multirate R-K and mis) compiling such a table to find the maximum possible time step (or maximum region of stability) is actually an open research question
> 
> I was guessing the CFL is mostly determined by the method and is less sensitive to the experiment, but I could be wrong. I understand it could be hard, as I saw in the doc that the CFL is sometimes heuristically determined:) I was just thinking about having something so that the users would know e.g. we could use a larger CFL for `LSRK144NiegemannDiehlBusch` than for `LSRK54CarpenterKennedy`, but I don't know what would be the best way to show this.

It is determined by both the temporal method and the spatial method/resolution chosen. In this context, the experiment setup is the one that determines the spatial resolution, and, therefore also the CFL condition. We tried to explain in the text why one method allows a larger step size than the other.

szy21 commented:
I'm a bit confused why CFL is dependent on the spatial resolution? I thought Δ𝑡 is calculated from CFL and Δx, but CFL is mostly dependent on the time stepper? 

The explanation in the text is clear to me. I was just thinking if a user chooses to use a different method, e.g. `ARK2GiraldoKellyConstantinescu`, it would be helpful to know what CFL to use. Having said that, since you have an example for each method, maybe it's good enough.

valeriabarra commented:
> I'm a bit confused why CFL is dependent on the spatial resolution? I thought Δ𝑡 is calculated from CFL and Δx, but CFL is mostly dependent on the time stepper?
> 
> The explanation in the text is clear to me. I was just thinking if a user chooses to use a different method, e.g. `ARK2GiraldoKellyConstantinescu`, it would be helpful to know what CFL to use. Having said that, since you have an example for each method, maybe it's good enough.

As you said, the [CFL](https://en.wikipedia.org/wiki/Courant%E2%80%93Friedrichs%E2%80%93Lewy_condition) condition depends on both  Δ𝑡 and Δx <-- and the latter depends on the spatial resolution chosen in that particular experiment considered and varies according to the spatial discretization method chosen (e.g., Finite Differences Vs Finite Elements or Spectral Elements etc.) and your mesh (e.g. structured or unstructured). I hope this helps : )

valeriabarra commented:
bors try

valeriabarra commented:
@jiahe23 @szy21 I have introduced most of the suggestions proposed. 

If satisfied with the addressed comments, please approve. I will take care of reducing the number of commits then.

valeriabarra commented:
bors r+
